opentracing-api/src/main/java/io/opentracing/propagation/BinaryAdapter.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2016-2017 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.opentracing.propagation;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+/**
+ * BinaryAdapter is a built-in carrier for Tracer.inject() and Tracer.extract(). BinaryAdapter
+ * is backed up by either a ByteArrayInputStream or a ByteArrayOutputStream, depending
+ * on whether it's defined as inbound or outbound, respectively.
+ */
+final class BinaryAdapter implements Binary {
+    final ByteArrayInputStream inputStream;
+    final ByteArrayOutputStream outputStream;
+
+    /**
+     * Create an outbound BinaryAdapter.
+     */
+    BinaryAdapter() {
+        this.outputStream = new ByteArrayOutputStream();
+        this.inputStream = null;
+    }
+
+    /**
+     * Create an inbound BinaryAdapter backed by the specified byte array.
+     */
+    BinaryAdapter(byte[] b) {
+        this.inputStream = new ByteArrayInputStream(b);
+        this.outputStream = null;
+    }
+
+    public boolean isOutbound() {
+        return outputStream != null;
+    }
+
+    public boolean isInbound() {
+        return inputStream != null;
+    }
+
+    public void write(byte[] b) throws IOException {
+        if (outputStream == null)
+            throw new UnsupportedOperationException();
+
+        outputStream.write(b);
+    }
+
+    public int read(byte [] b) throws IOException {
+        if (inputStream == null)
+            throw new UnsupportedOperationException();
+
+        return inputStream.read(b);
+    }
+
```

Comment:
Limiting the methods of ```ByteArrayOutputStream``` and ```ByteArrayInputStream``` to only ```read(byte[] b)``` and ```write(byte[] b)``` may limit tracing implementation libraries. Suppose this class simply had two getters:

```java
public InputStream getInput() {
  if (inputStream == null) {
    throw new UnsupportedOperationException();
  }
  return inputStream;
}

public OutputStream getOutput() {
  if (outputStream == null) {
    throw new UnsupportedOperationException();
  }
  return outputStream;
}
```
This would allow much greater flexibility to the tracing implementations (by not being rigid on what/how to read and when to close) and yet satisfy the api as defined by ```inject``` and ```extract```.

Reply:
I personally think exposing the actual stream objects would be overloading the API, trying to cover all cases and scenarios eventually. In any case, if this is truly needed, can be added in the future.

Reply:
On a second thought: we *may* as well remove the write/read methods and end up exposing the streams too (after prototyping the `MockTracer` support for BINARY) ;)